BJS2-25764

Caveri95 opened the pull request:
- Создан метод, который достает из БД все объекты AnalyticsEvent с данным receiverId и eventType, сортирует их по новшеству. Работает как с параметром Interval, так и просто с промежетком по дате
- Добавлен обработчик ошибок
- Написаны тесты

Caveri95 committed:
BJS2-25764

- Создан метод, который достает из БД все объекты AnalyticsEvent с данным receiverId и eventType, сортирует их по новшеству. Работает как с параметром Interval, так и просто с промежетком по дате
- Добавлен обработчик ошибок
- Написаны тесты

Caveri95 committed:
BJS2-25779

- Добавлен эндпоинт на получение AnalyticsEvent по receiverId и временному диапазону

Caveri95 committed:
BJS2-25779

- Добавлен интеграционный тест на контроллер получения аналитики

Caveri95 committed:
BJS2-25779

- Добавлен тест валидатора

Caveri95 committed:
BJS2-25779

- Правки в тестах

pablo023 commented on src/main/java/faang/school/analytics/service/AnalyticsEventService.java:
```diff
@@ -1,19 +1,57 @@
+                .toList();
+    }
+
+    private boolean filterBetweenDate(LocalDateTime from, LocalDateTime to, LocalDateTime dateTime) {
+        return dateTime.isAfter(from) && dateTime.isBefore(to);
+    }
+
+    private boolean filterByIntervalDays(Interval interval, LocalDateTime dateTime) {
+        int daysByInterval = interval.getDaysByInterval(interval);
+        return dateTime.isAfter(LocalDateTime.now().minusDays(daysByInterval));
+    }
```
Эту фильтрацию легче и проще описать на уровне запроса к БД.

Caveri95 replied:
окей, сделаю метод в репозитории

Caveri95 replied:
если мы используем готовый метод из репозитория, который возвращает нам уже стрим из данных по id и eventType, то эти методы (фильтрации на уровне репозитория) я не смогу использовать на репозитории. Мне нужно работать с готовым стримом, поэтому сделал в коде именно обработку
![image](https://github.com/user-attachments/assets/31ec018b-8ad6-46c3-9f39-62b19b066459)

pablo023 replied:
Я имел ввиду написать вообще кастомную кверю и метод в репозитории. Если хочешь - попробуй сделать, было бы интересно, если нет - можешь мержить, и кидать в done

pablo023 commented on src/main/java/faang/school/analytics/service/AnalyticsEventService.java:
```diff
@@ -1,19 +1,57 @@
+
+    @Transactional(readOnly = true)
+    public List<AnalyticEventDto> getAnalytics(long receiverId,
+                                               EventType eventType,
+                                               Interval interval,
+                                               LocalDateTime from,
+                                               LocalDateTime to) {
```
Это лучше собрать в дто

Caveri95 replied:
да, хорошо

pablo023 reviewed (changes requested):
Вынеси логику фильтрации в логику запроса к БД на репозиторном слое, чтобы мощностями БД это делать, а не сервисом.

Caveri95 committed:
BJS2-25764

- Добавлены два метода на поиск аналитике в БД по двум датам и пот одной дате
- Добавлен ДТО для запроса аналитики
- Правки в тестах

Caveri95 committed:
BJS2-25779

- Исправлен метод контроллера на Post, чтобы можно было передать тело запроса для получения аналитики
- Исправлены тесты
- Создан объект DTO для запроса аналитики

Caveri95 committed:
Merge branch 'titan-master-bc5-BJS2-25764' into titan-master-bc5-BJS2-25779

# Conflicts:
#	src/main/java/faang/school/analytics/dto/AnalyticInfoDto.java
#	src/main/java/faang/school/analytics/model/Interval.java
#	src/main/java/faang/school/analytics/service/AnalyticsEventService.java
#	src/test/java/faang/school/analytics/service/AnalyticsEventServiceTest.java

Caveri95 committed:
BJS2-25779

- conflict dix

Caveri95 committed:
Merge pull request #145 from CorporationX/titan-master-bc5-BJS2-25779

BJS2-25779
